### PR TITLE
feat(query-generator): Generate INSERTs using bind parameters

### PR DESCRIPTION
**Pull request by gazoakley**
### Pull Request check-list

_Please make sure to review and check all of these items:_

- [x] Does `npm run test` or `npm run test-DIALECT` pass with this change (including linting)?
- [x] Does the description below contain a link to an existing issue (Closes #[issue]) or a description of the issue you are solving?
- [x] Have you added new tests to prevent regressions?
- [x] Is a documentation update included (if this change modifies existing APIs, or introduces new ones)?
- [x] Did you follow the commit message conventions explained in [CONTRIBUTING.md](../CONTRIBUTING.md)?

### Description of change

This provides very basic support for using bind parameters when generating INSERT queries and partly addresses #1608. Re-do of #8923

### CODE REVIEW

File: test/unit/sql/insert.test.js [hunk 1]
```diff
@@ -31,8 +31,37 @@ describe(Support.getTestDialectTeaser('SQL'), () => {
...
+      const values = {
+        user: 'bindtest',
+        name: Support.sequelize.fn('UPPER', 'bindtest')
+      };
+      const options = {
+        bindParams: true
+      };
+      expectsql(sql.insertQuery(User.tableName, values, User.rawAttributes, options),
+        {
+          mssql: 'INSERT INTO [users] ([user],[name]) VALUES ($1,UPPER(N\'bindtest\'));',
```
Comment (janmeier) [inline test/unit/sql/insert.test.js]: Ideally I'd like to assert on bind parameters as well. Something along the lines of 

```
expectsql(sql..., {
 query: {
 ....
  }.
  bind: [...]
})
```

**Comment by gazoakley** [comment]
Thanks @janmeier . I've put in a flag out of cautiousness - it also means anyone enabling the flag will be aware of the limitations of the current implementation. It could be enabled by default in a future version.

File: test/unit/sql/insert.test.js (hunk 1 above)
Comment (gazoakley) [reply test/unit/sql/insert.test.js]: Good call - I'll do some tweaks shortly

**Comment by janmeier** [comment]
@sequelize/orm WDYT? Should we hide this behind a flag?

**Comment by janmeier** [comment]
What are the limitations exactly? Of the top of my head I can't come up with any other than functions that might cause problems for inserts and you are already handling that

**Comment by gazoakley** [comment]
Whilst I've named the option/flag `bindParams` it currently only supports bind parameters for non-bulk INSERTs - I've documented this as a limitation here:

https://github.com/sequelize/sequelize/pull/9371/files#diff-4be2078a125fe364f7198675d51bc688R123

```
// Generate queries with bind parameters
// Currently only has an effect for non-bulk INSERTs
// - default: false
bindParams: true
```

Merging this won't add support for bind parameters to everywhere in the query generator, but having this behind a documented option gives users the benefit of bind parameters whilst ensuring they are aware of where it is supported. There's always a possibility to enable this option by default in the future.

I don't think there should be a need to rerun the whole integration test suite twice - there are specific tests that target this option.

**Comment by janmeier** [comment]
I think what Erik is getting at is, that you remove the flag and run the full integration test suite with all its inserts again the new could we would have no problems releasing this without a flag.
If it can run against all our integration tests I'm sure 99% of cases are covered :)

**Comment by gazoakley** [comment]
At a brief glance 67 unit tests need updating to use the new format across the different dialects - I haven't looked at updating the integration tests yet. I'll see if I can get some time to make the changes.

**Comment by gazoakley** [comment]
@janmeier I'd come to the same conclusion. I'm thinking of passing a function as a third param to the `_stringify` methods that returns a `$n` placeholder and updates the bind params array.

**Comment by janmeier** [comment]
> @janmeier I'd come to the same conclusion. I'm thinking of passing a function as a third param to the _stringify methods that returns a $n placeholder and updates the bind params array.

Sounds like a good idea! Ping me when you have something ready for review (or let me know if you need my help)

**Comment by sushantdhiman** [comment]
Closing this PR as work is now continued in https://github.com/sequelize/sequelize/pull/9431

**Pull request closed without merging**